File: services/api/app/features/search/schemas.py
from datetime import date

from app.domain import DocumentType
from app.schemas import CamelModel


class ClinicalSearchRequest(CamelModel):
    query: str
    document_types: list[DocumentType] | None = None
    limit: int | None = None


class PatientSummary(CamelModel):
    id: str
    display_name: str


class BestMatch(CamelModel):
    document_id: str
    document_type: DocumentType
    document_title: str
    document_date: date
    snippet: str
    relevance_score: float


class SearchResult(CamelModel):
    patient: PatientSummary
    best_match: BestMatch
    additional_matching_documents: int


class SearchMeta(CamelModel):
    result_count: int
    took_ms: int


class ClinicalSearchResponse(CamelModel):
    query: str
    results: list[SearchResult]
    meta: SearchMeta

File: apps/web/features/search/schemas.ts
import { z } from "zod";

import { DocumentTypeSchema } from "@/features/clinical-documents/document-types";

export const ClinicalSearchRequestSchema = z.object({
  query: z.string(),
  documentTypes: z.array(DocumentTypeSchema).optional(),
  limit: z.number().int().positive().optional(),
});

export const PatientSummarySchema = z.object({
  id: z.string(),
  displayName: z.string(),
});

export const BestMatchSchema = z.object({
  documentId: z.string(),
  documentType: DocumentTypeSchema,
  documentTitle: z.string(),
  documentDate: z.string(),
  snippet: z.string(),
  relevanceScore: z.number(),
});

export const SearchResultSchema = z.object({
  patient: PatientSummarySchema,
  bestMatch: BestMatchSchema,
  additionalMatchingDocuments: z.number().int().nonnegative(),
});

export const ClinicalSearchResponseSchema = z.object({
  query: z.string(),
  results: z.array(SearchResultSchema),
  meta: z.object({
    resultCount: z.number().int().nonnegative(),
    tookMs: z.number().int().nonnegative(),
  }),
});

export type ClinicalSearchRequest = z.infer<typeof ClinicalSearchRequestSchema>;
export type ClinicalSearchResponse = z.infer<typeof ClinicalSearchResponseSchema>;
